app/util/cron.py

```diff
@@ -0,0 +1,69 @@
+import atexit
+import datetime
+import json
+import time
+
+from app import app
+from sqlalchemy import func, exc
+from app.model.email import Email, EmailStatus
+from app.util.db import db
+from app.util.mailer import mailer
+from apscheduler.schedulers.background import BackgroundScheduler
+from apscheduler.triggers.interval import IntervalTrigger
+
+JOB_INTERVAL_SEC = 10
+RETRY_INTERVAL_MIN = 10
+
+
+def send_mail():
+    with app.app_context():
+        mails_new = Email.query \
+            .filter(
+            Email.status == False,
+            Email.counter == 0
+        ).all()
+        mails_failed = Email.query \
+            .filter(
+            Email.status == False,
+            Email.last_sent_at < (datetime.datetime.utcnow() - datetime.timedelta(minutes=RETRY_INTERVAL_MIN)),
+            Email.counter < 4
+        ).all()
+
+        for mail_message in mails_new + mails_failed:
+            app.logger.info("Sending email #" + str(mail_message.id))
+            recipients = [mail_message.mail_to]
+            response = {}
+            if mail_message.mail_cc is not None and mail_message.mail_cc:
+                recipients += [mail_message.mail_cc]
+            if mail_message.mail_bcc is not None and mail_message.mail_bcc:
+                recipients += [mail_message.mail_bcc]
+
+            try:
+                response = mailer.send_message(mail_message.mail_from, recipients, mail_message.body)
+            except Exception as e:
+                app.logger.error(str(e))
+
+            if len(response) > 0:
+                app.logger.error('response: ' + response)
+
+            try:
+                mail_message.last_sent_at = datetime.datetime.utcnow()
+                mail_message.counter = mail_message.counter + 1
+                mail_message.status = False if len(response) > 0 else True
```

Comment:
If mailer returns any response it is treated as an error. Is really so?

Reply:
yes, according smtplib doc the sendmail function will return an empty dictionary on success. In case of failure it returns a dictionary, with one entry for each recipient that was refused.